File: java/tasks/io/File.java
package tasks.io;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

/** An interface for interfacing with file system files. */
public interface File {

  /** Opens an input stream to read the contents of the file. */
  InputStream openInputStream();

  /** Opens an output stream to write to the contents of the file. */
  OutputStream openOutputStream();

  static File fromPath(String path) {
    return new File() {

      @Override
      public InputStream openInputStream() {
        try {
          return new FileInputStream(path);
        } catch (IOException ex) {
          throw new RuntimeException(ex);
        }
      }

      @Override
      public OutputStream openOutputStream() {
        try {
          return new FileOutputStream(path);
        } catch (IOException ex) {
          throw new RuntimeException(ex);
        }
      }
    };
  }
}

File: src/tasks/io/File.kt
package tasks.io

import java.io.FileNotFoundException
import java.io.FileReader
import java.io.FileWriter
import java.io.IOException
import java.io.Reader
import java.io.StringReader
import java.io.Writer

/** An interface for interfacing with file system files.  */
interface File {
  fun openWriter(): Writer
  fun openReader(): Reader

  companion object {
    fun fromPath(path: String): File {
      return object : File {
        override fun openReader(): Reader {
          return try {
            FileReader(path)
          } catch (ex: FileNotFoundException) {
            StringReader("")
          }
        }

        override fun openWriter(): Writer {
          return try {
            FileWriter(path)
          } catch (ex: IOException) {
            throw RuntimeException(ex)
          }
        }
      }
    }
  }
}